### PR TITLE
Reset _isLocked property prior to it being determined (fix for #164)

### LINKED ISSUE #164: toggling _isAvailable can result in inappropriate locking
### Subject of the issue/enhancement/features
A custom plugin determines which articles are available based on some criteria. Each of these articles are configured to use trickle. The trickle plugin will set the `_isLocked` property on the articles and blocks as expected. At some later time all but one of the mentioned articles are set to `_isAvailable:false`. The article that is set to `_isAvailable:true` may still retain `_isLocked:true`. Rendering therefore stops at the article and the first block is not rendered as it should be.

Executing `applyLocks` does not correct the `_isLocked` property. Therefore, it is proposed that prior to `applyLocks` determining which models are locked, the `_isLocked` property is set to `false` on all models.

### Your environment
* all frameworks, devices and browsers

### Steps to reproduce
See above

### Expected behaviour
See above

### Actual behaviour
See above

### PR DESCRIPTION
**Pull request by chris-steele**
Resolves #164

### CODE REVIEW

File: js/models.js [hunk 1]
```diff
@@ -198,6 +198,7 @@ export function applyLocks() {
   const TrickleButtonModel = components.getModelClass('trickle-button');
   // Check all models for trickle potential
   Adapt.course.getAllDescendantModels(true).filter(model => model.get('_isAvailable')).forEach(siteModel => {
+    sideModel.set('_isLocked', false);
```
Comment (cahirodoherty-learningpool) [inline js/models.js:201]: Typo here
```suggestion
    siteModel.set('_isLocked', false);
```

Comment (chris-steele) [reply js/models.js]: crikey, what a blunder! good spot!

**Review by cahirodoherty-learningpool — APPROVED** [review]
👀

**Comment by oliverfoster** [comment]
I don't understand how this is a resolution to the issue. You've provided no explanation as to how it works. It's especially odd as the `_isAvailable: false` models are filters from the changes before this code block is executed and if the lock has changed it'll be applied in the second block of code at the bottom.

https://github.com/adaptlearning/adapt-contrib-trickle/blob/6e55edee0dc475f9f972bd45e9ae2f9905af4281/js/models.js#L200

https://github.com/adaptlearning/adapt-contrib-trickle/blob/6e55edee0dc475f9f972bd45e9ae2f9905af4281/js/models.js#L231-L237

**Review by oliverfoster — CHANGES_REQUESTED** [review]

File: js/models.js [hunk 2]
```diff
@@ -198,6 +198,9 @@ export function applyLocks() {
   const TrickleButtonModel = components.getModelClass('trickle-button');
   // Check all models for trickle potential
   Adapt.course.getAllDescendantModels(true).filter(model => model.get('_isAvailable')).forEach(siteModel => {
+    if (siteModel.has('_trickle') && siteModel.get('_trickle')._isEnabled) {
```
[the remark this answers is not in the document]
Comment (chris-steele) [reply js/models.js:201]: I'm pretty sure when I was debugging this that `const trickleConfig = getModelConfig(siteModel);` returns `undefined` which means the function returns on the following line and the article will not be unlocked. This is why I put the statement at the top of the function.

**Comment by chris-steele** [comment]
I want to proffer a similar but alternative solution in the next day or so

**Comment by chris-steele** [comment]
Just to follow up on this. The solution I've used is to reset the `_isLocked` status as follows in `applyLocks` prior to determining _isLocked:
```
Adapt.course.getAllDescendantModels(true).forEach(siteModel => {
    if (siteModel.get('_isTrickled')) {
      siteModel.setOnChildren('_isLocked', false);
    }
  });
```
I'm not sure if this is ideal, but I've had success with it where `_isAvailable` is dynamically set on articles.

**Comment by oliverfoster** [comment]
Any news on this? @chris-steele

**Comment by chris-steele** [comment]
Hey @oliverfoster the solution [posted ](https://github.com/adaptlearning/adapt-contrib-trickle/pull/165#issuecomment-1258012396) last week is the one I've used in several projects lately where there is dynamic switching of `_isAvailability`. It has worked a charm to be honest.

**Pull request closed without merging**